File: csharp/WorkOOP/Circle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WorkOOP
{
    // Реализация класса Circle – окружность.
    internal class Circle: IShape
    {
        protected Point2D p;    // p – точка, центр окружности.
        protected double r;     // r – радиус окружности.

        /*
         * Circle – конструктор, описывающий инициализацию объекта. На вход принимает центр окружности и его радиус.
         */
        public Circle(Point2D p, double r)
        {
            this.p = p;
            this.r = r;
        }

        /*
         * Реализация всех функций из интерфейса IShape
         */
        public Point2D getP() => p;
        public double getR() => r;
        public void setP(Point2D p) => this.p = p;
        public void setR(double r) => this.r = r;


        public double square() => Math.PI * r * r;
        public double length() => 2 * Math.PI * r;
        public IShape shift(Point2D a)
        {
            Point point = Point.add(a, p);
            p = new Point2D(point.getX());
            return this;
        }
        public IShape rot(double phi) => this;

        public IShape symAxis(int i)
        {
            p.symAxis(i);
            return this;
        }

        public bool cross(IShape i)
        {
            if (i.GetType() == typeof(Circle))
            {
                if (Point2D.sub(getP(), (i as Circle).getP()).abs() <= getR() + (i as Circle).getR())
                    return true;
                else
                    return false;
            }
            else
                return i.cross(this);            
        }
        public String toString() => "Center: " + p.toString()    + ", Radius: " + r;
    }
}

File: python/src/core/circle.py
from core.IShape import IShape
from core.point import Point
from core.point2d import Point2D

from math import pi

class Circle(IShape):
	def __init__(self, p, r):
		self.p = p
		self.r = r

	def __str__(self):
		return f"Circle: center: {self.p}; radius: {self.r}"

	def getR(self):
		return self.r

	def setR(self, newR):
		result = Circle(self.p, self.r)
		result.r = newR
		return result

	def getP(self):
		return self.p

	def setP(self, newP):
		result = Circle(self.p, self.r)
		result.p = newP
		return result

	def square(self):
		return pi * self.r * self.r

	def length(self):
		return 2 * pi * self.r

	def shift(self, point):
		result = Circle(self.p, self.r)
		result.p = result.p.add(point)
		return result

	def rot(self, phi):
		result = Circle(self.p, self.r)
		result.p = result.p.rot(phi)
		return result

	def symAxis(self, i):
		if i < 0:
			raise Exception("Ошибка: отрицательное пространство")
		if self.p.dim < i:
			raise Exception("Ошибка: имеет пространство меньшее, чем нужное")
		result = Circle(self.p, self.r)
		result.p = result.p.symAxis(i)
		return result

	def cross(self, figure):
		if figure.__class__.__name__ == "Circle":
			if Point2D.sub(self.getP(), figure.getP()).abs() <= self.getR() + figure.getR():
				return True
			else:
				return False
		return False